File: ferr-app/ui/js/components/progress.js
// progress.js — progress overlay + machine-readable line parser

const Progress = (() => {
    const overlay  = document.getElementById('progress-overlay');
    let _onCancel  = null;

    function show(title, opts = {}) {
        _onCancel = opts.onCancel ?? null;
        overlay.innerHTML = `
            <div class="progress-title">${title}</div>
            <div class="progress-filename" id="prog-filename">${t('preparing')}</div>
            <div class="progress-bar-track">
                <div class="progress-bar-fill" id="prog-bar" style="width:0%"></div>
            </div>
            <div class="progress-stats">
                <span id="prog-files">—</span>
                <span id="prog-bytes">—</span>
                <span id="prog-speed">—</span>
            </div>
            ${_onCancel ? `<button class="btn progress-cancel" id="prog-cancel">${t('cancel_btn')}</button>` : ''}
        `;
        if (_onCancel) {
            overlay.querySelector('#prog-cancel').addEventListener('click', () => {
                _onCancel?.();
                hide();
            });
        }
        overlay.classList.remove('hidden');
    }

    function hide() {
        overlay.classList.add('hidden');
        overlay.innerHTML = '';
        _onCancel = null;
        _buffer = '';
    }

    let _buffer = '';

    function update(chunk) {
        _buffer += chunk;
        const lines = _buffer.split('\n');
        _buffer = lines.pop() || '';

        for (const line of lines) {
            const ev = parseLine(line.trim());
            if (!ev) continue;

            if (ev.type === 'PROGRESS') {
            const pct = ev.totalBytes ? (ev.bytes / ev.totalBytes * 100).toFixed(1) : 0;
            _set('prog-bar', el => el.style.width = pct + '%');
            _set('prog-filename', el => el.textContent = ev.filename);
            _set('prog-files', el => el.textContent = `${ev.files}/${ev.totalFiles} files`);
            _set('prog-bytes', el => el.textContent = Fmt.bytes(ev.bytes));
            _set('prog-speed', el => el.textContent = ev.speed);
        }

        if (ev.type === 'SCAN_PROGRESS') {
            const pct = ev.total ? (ev.scanned / ev.total * 100).toFixed(1) : 0;
            _set('prog-bar', el => el.style.width = pct + '%');
            _set('prog-filename', el => el.textContent = ev.file);
            _set('prog-files', el => el.textContent = `${ev.scanned}/${ev.total} files`);
        }

            if (ev.type === 'COMPLETE') {
                _set('prog-bar', el => el.style.width = '100%');
                _set('prog-filename', el => el.textContent = typeof t === 'function' ? t('complete') : 'Complete');
            }
        }
    }

    function _set(id, fn) {
        const el = document.getElementById(id);
        if (el) fn(el);
    }

    // ── Machine-readable line parser ───────────────────────────────────────
    function parseLine(line) {
        if (!line) return null;

        if (line.startsWith('PROGRESS:')) {
            const rest = line.slice(9);
            const parts = rest.split('|');
            if (parts.length < 4) return null;
            const [bytes, totalBytes] = parts[0].split('/').map(Number);
            const [files, totalFiles] = parts[1].split('/').map(Number);
            return { type: 'PROGRESS', bytes, totalBytes, files, totalFiles, speed: parts[2], filename: parts[3] };
        }

        if (line.startsWith('COMPLETE:')) {
            const p = line.slice(9).split('|');
            return { type: 'COMPLETE', files: +p[0], bytes: +p[1], errors: +p[2], manifestPath: p[3] };
        }

        if (line.startsWith('ERROR:')) {
            const p = line.slice(6).split('|');
            return { type: 'ERROR', filename: p[0], kind: p[1], details: p[2] };
        }

        if (line.startsWith('SCAN_PROGRESS:')) {
            const [counts, file] = line.slice(14).split('|');
            const [scanned, total] = counts.split('/').map(Number);
            return { type: 'SCAN_PROGRESS', scanned, total, file };
        }

        if (line.startsWith('SCAN_RESULT:')) {
            const p = line.slice(12).split('|');
            return { type: 'SCAN_RESULT', ok: p[0] === 'OK', checked: +p[1], corrupted: +p[2] };
        }

        if (line.startsWith('VERIFY_RESULT:')) {
            const p = line.slice(14).split('|');
            return { type: 'VERIFY_RESULT', ok: p[0] === 'OK', matched: +p[1], mismatched: +p[2], missing: +p[3] };
        }

        if (line.startsWith('WATCH_DETECTED:')) {
            const [volume, mount] = line.slice(15).split('|');
            return { type: 'WATCH_DETECTED', volume, mount };
        }

        if (line.startsWith('WATCH_COPY_START:')) {
            const [volume, dest] = line.slice(17).split('|');
            return { type: 'WATCH_COPY_START', volume, dest };
        }

        return { type: 'UNKNOWN', raw: line };
    }

    return { show, hide, update, parseLine };
})();

File: ferr-app/src/progress.rs
#![allow(dead_code)]
// progress.rs — parse machine-readable ferr progress lines
//
// ferr outputs structured lines when --progress-format machine is passed:
//   PROGRESS:<bytes>/<total>|<files>/<total>|<speed>|<filename>
//   COMPLETE:<files>|<bytes>|<errors>|<manifest_path>
//   ERROR:<filename>|<type>|<details>
//   SCAN_PROGRESS:<scanned>/<total>|<file>
//   SCAN_RESULT:OK|<checked>|<corrupted>
//   VERIFY_RESULT:OK|<matched>|<mismatched>|<missing>
//   WATCH_DETECTED:<volume>|<mount>
//   WATCH_COPY_START:<volume>|<dest>

use serde::Serialize;

#[derive(Debug, Serialize)]
#[serde(tag = "type", rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ProgressEvent {
    Progress {
        bytes: u64,
        total_bytes: u64,
        files: u64,
        total_files: u64,
        speed: String,
        filename: String,
    },
    Complete {
        files: u64,
        bytes: u64,
        errors: u64,
        manifest_path: String,
    },
    Error {
        filename: String,
        kind: String,
        details: String,
    },
    ScanProgress {
        scanned: u64,
        total: u64,
        file: String,
    },
    ScanResult {
        ok: bool,
        checked: u64,
        corrupted: u64,
    },
    VerifyResult {
        ok: bool,
        matched: u64,
        mismatched: u64,
        missing: u64,
    },
    WatchDetected {
        volume: String,
        mount: String,
    },
    WatchCopyStart {
        volume: String,
        dest: String,
    },
    Unknown {
        raw: String,
    },
}

pub fn parse(line: &str) -> ProgressEvent {
    if let Some(rest) = line.strip_prefix("PROGRESS:") {
        let parts: Vec<&str> = rest.splitn(4, '|').collect();
        if parts.len() == 4 {
            let (bytes, total_bytes) = split_slash(parts[0]);
            let (files, total_files) = split_slash(parts[1]);
            return ProgressEvent::Progress {
                bytes,
                total_bytes,
                files,
                total_files,
                speed: parts[2].to_string(),
                filename: parts[3].to_string(),
            };
        }
    }

    if let Some(rest) = line.strip_prefix("COMPLETE:") {
        let p: Vec<&str> = rest.splitn(4, '|').collect();
        if p.len() == 4 {
            return ProgressEvent::Complete {
                files: p[0].parse().unwrap_or(0),
                bytes: p[1].parse().unwrap_or(0),
                errors: p[2].parse().unwrap_or(0),
                manifest_path: p[3].to_string(),
            };
        }
    }

    if let Some(rest) = line.strip_prefix("ERROR:") {
        let p: Vec<&str> = rest.splitn(3, '|').collect();
        if p.len() == 3 {
            return ProgressEvent::Error {
                filename: p[0].to_string(),
                kind: p[1].to_string(),
                details: p[2].to_string(),
            };
        }
    }

    if let Some(rest) = line.strip_prefix("SCAN_PROGRESS:") {
        let p: Vec<&str> = rest.splitn(2, '|').collect();
        if p.len() == 2 {
            let (scanned, total) = split_slash(p[0]);
            return ProgressEvent::ScanProgress {
                scanned,
                total,
                file: p[1].to_string(),
            };
        }
    }

    if let Some(rest) = line.strip_prefix("SCAN_RESULT:") {
        let p: Vec<&str> = rest.splitn(3, '|').collect();
        if p.len() == 3 {
            return ProgressEvent::ScanResult {
                ok: p[0] == "OK",
                checked: p[1].parse().unwrap_or(0),
                corrupted: p[2].parse().unwrap_or(0),
            };
        }
    }

    if let Some(rest) = line.strip_prefix("VERIFY_RESULT:") {
        let p: Vec<&str> = rest.splitn(4, '|').collect();
        if p.len() == 4 {
            return ProgressEvent::VerifyResult {
                ok: p[0] == "OK",
                matched: p[1].parse().unwrap_or(0),
                mismatched: p[2].parse().unwrap_or(0),
                missing: p[3].parse().unwrap_or(0),
            };
        }
    }

    if let Some(rest) = line.strip_prefix("WATCH_DETECTED:") {
        let p: Vec<&str> = rest.splitn(2, '|').collect();
        if p.len() == 2 {
            return ProgressEvent::WatchDetected {
                volume: p[0].to_string(),
                mount: p[1].to_string(),
            };
        }
    }

    if let Some(rest) = line.strip_prefix("WATCH_COPY_START:") {
        let p: Vec<&str> = rest.splitn(2, '|').collect();
        if p.len() == 2 {
            return ProgressEvent::WatchCopyStart {
                volume: p[0].to_string(),
                dest: p[1].to_string(),
            };
        }
    }

    ProgressEvent::Unknown { raw: line.to_string() }
}

fn split_slash(s: &str) -> (u64, u64) {
    let mut it = s.splitn(2, '/');
    let a = it.next().unwrap_or("0").parse().unwrap_or(0);
    let b = it.next().unwrap_or("0").parse().unwrap_or(0);
    (a, b)
}
